### PR TITLE
core: handle scoped keywords if flags are not allowed

### PR DESCRIPTION
**Pull request by aurelien-brabant**
I was playing with Hyprlang to handle configuration for a project of mine and found out that keywords inside categories are not properly handled (they are simply being ignored).

I added tests to make sure only the correctly scoped keywords are calling the handler.

### CODE REVIEW

File: tests/config/config.conf [hunk 1]
```diff
@@ -40,14 +42,20 @@ testCategory {
         testValueNest = 1
         nested2 {
             testValueNest = 1
-        }
-    }
+    	}
+		categoryKeyword = this one should not either
+	}
+
+	categoryKeyword = we are having fun
```
Comment (vaxerski) [inline tests/config/config.conf:49]: don't mix tabs and spaces please it messes up rendering :P

we use spaces

File: tests/config/config.conf [hunk 2]
```diff
@@ -28,6 +28,8 @@ errorVariable = true
 
 # hyprlang noerror false
 
+categoryKeyword = oops, this one shouldn't call the handler, not fun
```
Comment (vaxerski) [inline tests/config/config.conf:31]: if it's a category one, this should throw an error, no?

File: src/config.cpp [hunk 3]
```diff
@@ -563,9 +563,30 @@ CParseResult CConfig::parseLine(std::string line, bool dynamic) {
...
+
+                    if (actual != impl->categories[depth]) {
+                        break;
+                    }
+
+                    idx = colon + 1;
+                    ++depth;
+                }
+
+                if (depth != impl->categories.size() || h.name.substr(idx) != LHS) {
```
Comment (vaxerski) [inline src/config.cpp:586]: no {} around short ifs

File: tests/config/config.conf (hunk 2 above)
Comment (aurelien-brabant) [reply tests/config/config.conf:31]: Well, in my eyes `categoryKeyword` and `category::categoryKeyword` are effectively totally different keywords. The naming here might feel a little bit confusing but it's to ensure the handler for the latter is not called for the former as well. This one is intended to be a legit config option but that shouldn't be picked up by the handler.

Comment (vaxerski) [reply tests/config/config.conf:31]: oh, right. Missed that, because both variables are missing a test case :P

File: tests/parse/main.cpp [hunk 4]
```diff
@@ -162,6 +174,10 @@ int main(int argc, char** argv, char** envp) {
         EXPECT(barrelRoll, true);
         EXPECT(flagsFound, std::string{"abc"});
 
+        EXPECT(categoryKeywordActualValues[0], "we are having fun");
```
Comment (vaxerski) [inline tests/parse/main.cpp:177]: you should use .at(). If for whatever reason the handler fails, `[1]` will produce an access out-of-bounds, while .at will throw.

**Comment by vaxerski** [comment]
also rebase on main to fix ci

**Review by vaxerski — APPROVED** [review]
thanks!

**Comment by fufexan** [comment]
@vaxerski you forgot to merge.

**Comment by vaxerski** [comment]
no I totally didnt

**Comment by zjeffer** [comment]
After updating my whole system with `paru`, this config stopped working:

```
animations {
  enabled = true

  # Some default animations, see https://wiki.hyprland.org/Configuring/Animations/ for more
  animation = windows, 1, 1, default
  animation = windowsOut, 1, 2, default
  animation = border, 1, 5, default
  animation = borderangle, 1, 8, default
  animation = fade, 1, 3, default
  animation = workspaces, 1, 4, default
  animation = specialWorkspace, 1, 4, default, slidevert
}
```

I get the config error `config option <animations.animation> does not exist`.

Instead, I now have to do this:

```
animations {
  enabled = true
}
# Some default animations, see https://wiki.hyprland.org/Configuring/Animations/ for more
animation = windows, 1, 1, default
animation = windowsOut, 1, 2, default
animation = border, 1, 5, default
animation = borderangle, 1, 8, default
animation = fade, 1, 3, default
animation = workspaces, 1, 4, default
animation = specialWorkspace, 1, 4, default, slidevert
```

I believe this commit is responsible.

**Comment by vaxerski** [comment]
fixed and added a test case: 5df0174fd09de4ac5475233d65ffc703e89b82eb

if you want your keyword to only work in global scope (not in categories) prefix it with `:`

`:globalScopeOnly`
`anyScope`
`category:onlyThisCat`

**Pull request merged**